Adding Python Container Documentation (#3437)
microsoft/vscode-docs: docs/containers/quickstart-python.md

## Prerequisites

- Docker Desktop and the VS Code Docker extension must be installed as described in the [overview](overview.md#installation).
- For Python development, complete all [Getting started with Python](/docs/python/python-tutorial.md) steps
- A runnable Python application

### Create a Python project

If you don't have a Python project already, follow these commands sequentially from the terminal:

```bash
pip install django
django-admin startproject helloworld
cd helloworld
code .
```

If you want to containerize a complete Django or Flask web app, you can use one of the following samples:

- [python-sample-vscode-django-tutorial](https://github.com/microsoft/python-sample-vscode-django-tutorial/tree/tutorial), which is the result of following the [Django Tutorial](/docs/python/tutorial-django.md)

- [python-sample-vscode-flask-tutorial](https://github.com/microsoft/python-sample-vscode-flask-tutorial/tree/tutorial), which is the result of following the [Flask Tutorial](/docs/python/tutorial-flask.md)

After verifying your app runs properly, you can now Dockerize your application.

## Add Docker files to the project

1. Open the project folder in VS Code.

> 2. Open the Command Palette (`kb(workbench.action.showCommands)`) and use **Docker: Add Docker Files to Workspace...** command:

Reviewer: We usually use the numeral 1 for all list items - that way you can add/substract items over time and the published list remains ordered without any manual work.

Author: Yes, this is something I wanted to do, however for some reason with putting the **Tip**: in between 5 and 6, it breaks the numbering. I know this is not ideal, but I do not know a valid alternative

Author: But I did go ahead and update any other list where possible to be all 1s. Thanks for that suggestion

Reviewer: If you indent the Tip, so that the beginning of that block aligns with the first text in the previous number, it will be a child of the item and won't break the list.

    ![Add Dockerfile to a Python project](images/quickstarts/python-add-python.png)

3. When the prompt appears, select **Python: Django**, **Python: Flask**, or **Python: General** as the app type. For this tutorial, we will select **Python: Django**.
4. Select either **Yes** or **No** when prompted to include [Docker Compose](https://docs.docker.com/compose/) files.

> 5. Enter the relative path to the app’s entry point. This excludes the workspace folder you start from. According to [official Django documentation](https://docs.djangoproject.com/en/3.0/intro/tutorial01/#creating-a-project), this path is commonly `manage.py` (root folder) or `subfolder_name/manage.py`. According to [official Flask documentation](https://flask.palletsprojects.com/en/1.1.x/api/), this is the path to where you create your Flask instance.

Reviewer: ```suggestion
5. Enter the relative path to the app's entry point. This excludes the workspace folder you start from. According to [official Django documentation](https://docs.djangoproject.com/en/3.0/intro/tutorial01/#creating-a-project), this path is commonly `manage.py` (root folder) or `subfolder_name/manage.py`. According to [official Flask documentation](https://flask.palletsprojects.com/en/1.1.x/api/), this is the path to where you create your Flask instance.
```

>**Tip**: You may also enter the path to a folder name as long as this folder includes a `__main__.py` file.

> 6. If **Python: Django** or **Python: Flask** was selected, specify app port for local development. Django defaults to port 8000 while Flask defaults to port 5000, however, any unused port will work.

Reviewer: ```suggestion
6. If **Python: Django** or **Python: Flask** was selected, specify app port for local development. Django defaults to port 8000 while Flask defaults to port 5000; however, any unused port will work.
```

    - A `Dockerfile`. To learn more about Intellisense in this file, refer to the [overview](/docs/containers/overview.md).

    - A `.dockerignore` file to reduce the image size by excluding files and folders that aren't needed such as `.git`, `.vscode`, and `__pycache__`.

    - If Docker Compose was selected, a `docker-compose.yml` file.

    - If one does not already exist, a `requirements.txt` file for capturing all app dependencies.
    > **Tip**: If the virtual environment/host machine is meant to be identical to the container, ensure app dependencies are ported over by running `pip freeze > requirements.txt` in the terminal to overwrite this file.

### File modifications for Django/Flask apps

To give Python Web Developers a great starting point, we chose to use [Gunicorn](https://gunicorn.org/#docs) as the default web server. Since it is referenced in the default Dockerfile, it is included as a dependency in the `requirements.txt` file.

> **Note**: To use Gunicorn as your web server, it must be included in the `requirements.txt` file as an app dependency. It does not need to be installed in your virtual environment/host machine.

#### Django Apps

> To use Gunicorn, it must bind to an application callable as an entry point. This callable is declared in the `wsgi.py` file of a Django application. To accomplish this binding, the final line in the Dockerfile says:

Reviewer: Application callable - will this term be known to all users?  Could it be explained?  "it must bind to an application callable as an entry point'.  It seems like callable is being used as a noun, but I think of it as an adjective.

Author: It's a fairly used term in WSGI, but I changed it to "... must bind to an application callable (what the application server uses to communicate with your code) as an entry point." for better clarity. Thanks

```docker
CMD ["gunicorn", "--bind", "0.0.0.0:8000", "{workspace_folder_name}.wsgi"]`
```

This assumption is based off Django's default project structure. If your project does not follow `django_project_name (workspace folder) -> django_project_name (subfolder) -> wsgi.py`, you must overwrite the Gunicorn entry point in the Dockerfile to locate the correct `wsgi.py` file.

> > **Tip**: If your `wsgi.py` file is in the root folder, the final argument in the command above will be `"wsgi"`. Within subfolders, the argument would be `"subfolder1_name.subfolder2_name.wsgi"`.

Reviewer: ```suggestion
> **Tip**: If your `wsgi.py` file is in the root folder, the final argument in the command above will be `wsgi`. Within subfolders, the argument would be `subfolder1_name.subfolder2_name.wsgi`.
```

Author: In this suggestion, the user will actually need the double quotations "" around their words. Do you still believe I should change it?

Author: Personally I would like to keep it as is for clarity

Reviewer: Please keep it if they need it, sorry!

#### Flask Apps

> To use Gunicorn, it must bind to an application callable as an entry point. This callable corresponds with the **file location** and **variable name** of your created Flask instance. According to [official Flask Documentation](https://flask.palletsprojects.com/en/1.1.x/api/), users generally create a Flask instance in the main module or in the `__init__.py` file of their package in this manner:

Reviewer: here also - application callable

Author: Same change above. 👍

```python
from flask import Flask
app = Flask(__name__) # Flask instance named app
```

To accomplish this binding, the final line in the Dockerfile says:

```docker
CMD ["gunicorn", "--bind", "0.0.0.0:5000", "{subfolder}.{module_file}:app"]
```

> You have already given the path to the Flask instance, but the Docker extension assumes your Flask instance variable is named `app`. If this is not the case, the variable name must be overwritten in the Dockerfile.

Reviewer: I'd rewrite the passive voice in the second sentence here - do you mean "you must overwrite the variable name" or does the tool do something -it's ambiguous.

Author: Changed to "If this is not the case, you must change the variable name in the Dockerfile."

> > **Tip**: If your main module was in the root folder as a file named `main.py` and had a Flask instance variable was named `myapp`, the final argument in the command above will be `"main:myapp"`. Within subfolders, the argument would be `"subfolder1_name.subfolder2_name.main:myapp"`.

Reviewer: ```suggestion
> **Tip**: If your main module was in the root folder as a file named `main.py` and had a Flask instance variable was named `myapp`, the final argument in the command above will be `main:myapp`. Within subfolders, the argument would be `subfolder1_name.subfolder2_name.main:myapp`.
```

Author: Similar to the other suggestion, the user will actually need the double quotations "" around their words when they put it in the gunicorn command. Do you still believe I should change it?

Reviewer: No please do what's right, thanks.

## Build, run, and debug the container

The **Docker: Add Docker Files to Workspace...** command automatically creates a Docker task configuration to build and run your container in debug mode. To debug your Python app container:

1. Navigate to the `manage.py` file and set a breakpoint on this line:

```python
  os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'helloworld.settings')
```

>**Note**: If you have created an app project as shown in the [Create a Django app](https://code.visualstudio.com/docs/python/tutorial-django#_create-a-django-app) section of the Django tutorial, you may set a breakpoint in `views.py` or wherever you choose.

> 2. Navigate to Run and Debug then select **Docker: Python - Django**.

Reviewer: If these are UI elements "Run" and "Debug" - then put them in bold.

  ![Selected Docker debug configuration](images/quickstarts/python-debug-configuration.png)

3. Start debugging using the `kb(workbench.action.debug.start)` key.
    - The Docker image builds.
    - The Docker container runs.
    - The python debugger hits the breakpoint in `manage.py`.

> 4. Step over this line once

Reviewer: ```suggestion
4. Step over this line once.
```

> 5. Navigate to the Debug Console and type `os.environ["DJANGO_SETTINGS_MODULE"]`

Reviewer: ```suggestion
5. Navigate to the Debug Console and type `os.environ["DJANGO_SETTINGS_MODULE"]`.
```
